@Listwon I've just upgraded to SCons 4.0.1 now (without upgrading Python), I get the same error:
```
1>scons: Reading SConscript files ...
1>Build option 'custom_modules' must point to an existing directory.
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets(44,5): error MSB3073: The command "echo Starting SCons && cmd /V /C set "plat=x64" ^& (if "x64"=="x64" (set "plat=x86_amd64")) ^& set "tools=yes" ^& (if "debug"=="release" (set "tools=no")) ^& set "custom_modules=" ^& call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat" !plat! ^& scons --directory="D:\src\godot\3" platform=windows progress=no target=debug tools=!tools! custom_modules=!custom_modules! -j6" exited with code 255.
1>Done building project "godot.vcxproj" -- FAILED.
```

It would seem like a regression from SCons in fact, but it's difficult to blame the SCons fellows because of compatibility breakage with 4.0.1.

When printing `env_base["custom_modules"]` from within `SConstruct` the value is: `!custom_modules!`, because somehow it gets undefined. Perhaps there was some breaking changes which return `None` vs "" (empty string) for a construction environment (`env`) which doesn't exists.

> Isn't it handled already in custom.py?

That's certainly an option, and I'd recommend using that if you need to configure more advanced build options currently, it's just that it wouldn't be intuitive for the rest of the people I guess, and I feel like `custom_modules` is essential to prosperity of C++ development outside of Godot, for Godot.

VS still needs to define `PlatformTarget` for the target architecture, the rest can be configured outside of VS.

In any case, like I said, the current implementation is error-prone anyway, by now I've tested your implementation with and without `custom_modules`, both on SCons 3.1.1 and 4.0.1, so feel free to make the same changes on `master` first. Unless @akien-mga has some other ideas about this.

--- a/methods.py
+++ b/methods.py
@@ -572,13 +572,28 @@ def generate_vs_project(env, num_jobs):
             common_build_prefix = [
                 'cmd /V /C set "plat=$(PlatformTarget)"',
                 '(if "$(PlatformTarget)"=="x64" (set "plat=x86_amd64"))',
-                'set "tools=yes"',
+                'set "tools=%s"' % env["tools"],
                 '(if "$(Configuration)"=="release" (set "tools=no"))',
-                'set "custom_modules=%s"' % env["custom_modules"],
                 'call "' + batch_file + '" !plat!',
             ]
 
-            result = " ^& ".join(common_build_prefix + [commands])
+            # windows allows us to have spaces in paths, so we need
+            # to double quote off the directory. However, the path ends
+            # in a backslash, so we need to remove this, lest it escape the
+            # last double quote off, confusing MSBuild
+            common_build_postfix = [
+                "--directory=\"$(ProjectDir.TrimEnd('\\'))\"",
+                "platform=windows",
+                "target=$(Configuration)",
+                "progress=no",
+                "tools=!tools!",
+                "-j%s" % num_jobs,
+            ]
+
+            if env["custom_modules"]:
+                common_build_postfix.append("custom_modules=%s" % env["custom_modules"])
+
+            result = " ^& ".join(common_build_prefix + [" ".join([commands] + common_build_postfix)])
             return result
 
         env.AddToVSProject(env.core_sources)
@@ -588,22 +603,9 @@ def generate_vs_project(env, num_jobs):
         env.AddToVSProject(env.servers_sources)
         env.AddToVSProject(env.editor_sources)
 
-        # windows allows us to have spaces in paths, so we need
-        # to double quote off the directory. However, the path ends
-        # in a backslash, so we need to remove this, lest it escape the
-        # last double quote off, confusing MSBuild
-        env["MSVSBUILDCOM"] = build_commandline(
-            "scons --directory=\"$(ProjectDir.TrimEnd('\\'))\" platform=windows progress=no target=$(Configuration) tools=!tools! custom_modules=!custom_modules! -j"
-            + str(num_jobs)
-        )
-        env["MSVSREBUILDCOM"] = build_commandline(
-            "scons --directory=\"$(ProjectDir.TrimEnd('\\'))\" platform=windows progress=no target=$(Configuration) tools=!tools! vsproj=yes custom_modules=!custom_modules! -j"
-            + str(num_jobs)
-        )
-        env["MSVSCLEANCOM"] = build_commandline(
-            "scons --directory=\"$(ProjectDir.TrimEnd('\\'))\" --clean platform=windows progress=no target=$(Configuration) tools=!tools! custom_modules=!custom_modules! -j"
-            + str(num_jobs)
-        )
+        env["MSVSBUILDCOM"] = build_commandline("scons")
+        env["MSVSREBUILDCOM"] = build_commandline("scons vsproj=yes")
+        env["MSVSCLEANCOM"] = build_commandline("scons --clean")
 
         # This version information (Win32, x64, Debug, Release, Release_Debug seems to be
         # required for Visual Studio to understand that it needs to generate an NMAKE

